### DISCUSSION TITLE
Zigbee network size

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#970] [General]
This is just a question to understand what are the general usage of the zigpy libraries. Do you have information on how large are the zigbee network  size handle via zigpy libraries.

Is that an information available from the ZHA ?

I indeed see that the coordinators are capable of handling 200 devices, but I wonder how it works , how the coordinator behave/react with large network

**Comment by dmulcahey** [comment]
![CE545F85-51D1-421B-9FB5-869D8EE339AE](https://user-images.githubusercontent.com/1335687/164476369-f41068aa-2621-459b-95f6-e2aba5f56d92.jpeg)

I have > 150 devices and no issues. There are some HA users well north of 200.

**Reply by pipiche38** [reply]
Would be interested to know how many routers and type of routers you have?
How much traffic exists ? For instance do you have a lot of plug/switch which report instant power ?

And what type of coordinator do you run ?

**Reply by dmulcahey** [reply]
I’ll write some code later to get exact details when I have time. > 40 routers, somewhere near 20ish energy monitoring plugs / switches

**Reply by dmulcahey** [reply]
Coordinator is a Tube EZSP coordinator using https://www.silabs.com/documents/public/data-sheets/mgm12p-datasheet.pdf

**Reply by dmulcahey** [reply]
This network was previously running with a Nortek go control combo stick based on the EM358X chipset. It has not grown much since I switched over.

**Reply by Hedda** [reply]
Maybe an idea to request if non-intrusive/anonymized data about Zigbee devices can be added to opt-in Home Assistant Analytics?

https://www.home-assistant.io/blog/2021/11/12/100k-analytics/

https://analytics.home-assistant.io

Think it would be interesting to be able to dig just a little deeper into analytics for the [ZHA integration](https://www.home-assistant.io/integrations/zha) to see basic statistics for ex:

- radio type (e.i. deconz, ezsp, xbee, zigate, or znp)
- Zigbee network
  - Total amount of Zigbee devices
    - Amount of direct children (devices directly connected to coordinator)
    - Amount of routes
    - Amount of Zigbee Router devices
    - Amount of Zigbee End Device devices
    - Amount of non Zigbee 3.x devices

Maybe even a list top-10 brands of those Zigbee devices if brands of Zigbee devices used would not be considered too intrusive?

Home Assistant founders pointed out in the blog post that analytics can be used by developers and as an incentive to manufacturers.

**Comment by Hedda** [comment]
> I indeed see that the coordinators are capable of handling 200 devices, but I wonder how it works , how the coordinator behave/react with large network

As I understand the limitation of older Zigbee devices is much higher than newer Zigbee 3.0 devices due to Zigbee 3.0 security demanding much more resources. The total amount of Zigbee 3.0 devices therefor also depends on the Zigbee coordinator hardware due to CPU + RAM-memory resource constrictions, 

Meaning that some Zigbee coordinator hardware with faster CPU and more importantly more RAM-memory should normally be able to handle more Zigbee 3.0 devices, but still progarably limited at a around a maximum of around 200 Zigbee 3.0 devices, while both newer and older Zigbee coordinator hardware can handle many more older Zigbee 1.x devices (e.g. thousands of ZHA 1.x and  ZZL 1.x devices).

See ex. TI Z-Stack coordinator firmware limits -> https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/README.md

# Z-Stack coordinator firmwares
This directory contains various Z-Stack coordinator firmwares.

## Overiew
<table>
  <tr>
    <td><b>Z-Stack</b></td>
    <td><b>Device</b></td>
    <td><b>Zigbee</b></td>
    <td><b>Direct children</b></td>
    <td><b>Routes</b></td>
    <td><b>Notes</b></td>
  </tr>
  <tr>
    <td rowspan="2">Z-Stack_Home_1.2 (default)</td>
    <td>CC2531</td>
    <td>1.2 HA</td>
    <td>20</td>
    <td>30/0</td>
    <td></td>
  </tr>
  <tr>
    <td>CC2530, CC2530 + CC2591, CC2530 + CC2592</td>
    <td>1.2 HA</td>
    <td>16</td>
    <td>30/0</td>
    <td></td>
  </tr>
  <tr>
    <td>Z-Stack_Home_1.2 (source_routing)</td>
    <td>CC2531, CC2530, CC2530 + CC2591, CC2530 + CC2592</td>
    <td>1.2 HA</td>
    <td>5</td>
    <td>40/40</td>
    <td></td>
  </tr>
  <tr>
    <td rowspan="3">Z-Stack_3.0.x</td>
    <td>CC2531</td>
    <td>3.0</td>
    <td>15</td>
    <td>40/0</td>
    <td>
      - <a href="https://github.com/Koenkk/zigbee2mqtt/issues/1445">Discussion #1445</a>
      <br/>
      - Max 40 Zigbee 3.0 devices
    </td>
  </tr>
  <tr>
    <td>CC2530, CC2530 + CC2591, CC2530 + CC2592</td>
    <td>3.0</td>
    <td>10</td>
    <td>40/0</td>
    <td>
      - <a href="https://github.com/Koenkk/zigbee2mqtt/issues/1445">Discussion #1445</a>
      <br/>
      - Max 40 Zigbee 3.0 devices
    </td>
  </tr>
    <tr>
    <td>CC2538 + CC2592</td>
    <td>3.0</td>
    <td>100</td>
    <td>200/400</td>
    <td>
      - <a href="https://github.com/Koenkk/zigbee2mqtt/issues/1568">Discussion #1568</a>
      <br/>
      - Max 200 Zigbee 3.0 devices
    </td>
  </tr>
  <tr>
    <td rowspan="2">Z-Stack_3.x.0</td>
    <td>CC2652P, CC2652R, CC2652RB, CC1352P-2</td>
    <td>3.0</td>
    <td>50</td>
    <td>100/200</td>
    <td>
      - <a href="https://github.com/Koenkk/zigbee2mqtt/issues/1429">Discussion #1429</a>
      <br/>
      - Max 200 Zigbee 3.0 devices
    </td>
  </tr>
</table>

### I'm using a CC2530 or CC2531, which firmware should I use?
This depends:
- Zigbee 3.0 firmwares are **not** recommended for the CC2530 and CC2531 (since those are not powerful enough)
- If you have a network of 1 - 15 devices, the Z-Stack_Home_1.2 **default** firmware is recommended.
- If you have a network of 15+ devices, the Z-Stack_Home_1.2 **source routing** firmware is recommended.
- Note that the **source routing** firmware only supports 5 direct children, therefore you need to have routers in range of the coordinator.

### Legend
- *Z-Stack:* Z-Stack is the Zigbee stack for Texas Instruments Zigbee devices
- *Device:* The Zigbee hardware
- *Zigbee:* Zigbee version, for more information read [What's New in Zigbee 3.0](http://www.ti.com/lit/an/swra615a/swra615a.pdf)
- *Direct children:* Number of children that directly can join the coordinator. Note that this is **not** the maximum network size. For more information, read [the Zigbee2MQTT FAQ about device limit](http://www.zigbee2mqtt.io/information/FAQ.html#i-read-that-zigbee2mqtt-has-a-limit-of-15-devices-is-this-true)
- *Routes:* This indicates how many routes the coordinator can keep in memory. For example 100/200 would mean that: 100 normal routes and 200 source routes can be kept in memory. 0 source routes mean that source routing is disabled. Source routes improve the performance for larger (40+ nodes) networks. For more information read [Large ZigBee Networks and Source Routing
](http://cms.digi.com/resources/documentation/digidocs/90001537/references/r_large_zigbee_networks-source_routing.htm?TocPath=Working%20with%20Zigbee%7C_____14).

#### Notes
- Due to new security requirements in Zigbee 3.0, Zigbee 3.0 coordinators can only support a limited amount of Zigbee 3.0 devices. For Zigbee 1.2 coordinators there is no limit on the max number of Zigbee 3.0 devices that can join.

**Comment by MattWestb** [comment]
Zigbee 1.X using only the "well known" Zigbee09 key for manage the network and is not updating TC-link key (its not implanted) and the theoretical max devices is 64K devices.

Then building one Zigbee 3 network is moderate updating the TC-Link key and its need NVS for storing the updated keys.
Bellow is for the moment using hashed TC-Link keys so its only saving the network key and APS-Keys if some device is requesting some and i think its the same max devices in the network as Zigbee 1.X (= not knowing).

By running one coordinator in Zigbee 1.X mode all Zigbee 3 router device is being joined as no sleeping end device for not braking the comparability mode so having one no router network is little very bad !!

I think one old HUSBZB-1 or Telegesis EM357 can running one very large Zigbee network if using hashed TC-:Link keys without problems as long building the network in one good way.

Also is using lights its one limiting in the underlying IEEE 802.15.4 network that is 1.2 broadcast / seconds (I think its 9 packages in 8 seconds) and if doing more all routers is doping the broadcast for protecting the network for broadcast storms. 
And is using source routing the network manager is using broadcast for finding routers to not known route to devices that also must being in the calculation.

Joining device is also using broadcast commands that can getting problem if having much broadcast in the network.

One very bad was building the network is using HA (not ZHA) groups and all doing automatons that is truing lights on and off and changing its color and spamming the network with hundreds of unicast instead of one broadcast.

**Reply by Hedda** [reply]
> One very bad was building the network is using HA (not ZHA) groups and all doing automatons that is truing lights on and off and changing its color and spamming the network with hundreds of unicast instead of one broadcast.

Do you mean using groups in Home Assistant instead of using native Zigbee groups in ZHA? -> https://www.home-assistant.io/blog/2022/04/06/release-20224/

**Reply by MattWestb** [reply]
Working with lights in Ziogbee shall being done with Zigbee groups that is ZHA groups in HA.
HA groups is making ZHA sending all commands with unicast to all devices but its up to the user to doing it right.

**Reply by Hedda** [reply]
> Working with lights in Ziogbee shall being done with Zigbee groups that is ZHA groups in HA. HA groups is making ZHA sending all commands with unicast to all devices but its up to the user to doing it right.

So best would maybe be if the Home Assistant developers change their group implementation to make use of ZHA groups if and when the entities in that Home Assistant groups only consist of ZHA devices? https://www.home-assistant.io/blog/2022/04/06/release-20224/

**Reply by dmulcahey** [reply]
No that’s a separation of concerns thing…. The core should not have special logic in a group implementation for a specific integration. That’s why we have our own capabilities to create groups in ZHA
